### PR TITLE
Stormcast Customization

### PR DESCRIPTION
**Pull request by jleinonen**
# Modulus Pull Request

## Description

Modifies the StormCast code to be trainable with custom datasets, as well as adding a few additional features.

Overview of changes:
* Combine `.to` commands in trainer (device and dtype in one call)
* Save regression and diffusion checkpoints in different directories in rundir to avoid diffusion training trying to load regression checkpoint and vice versa
* List included HRRR channels in dataset config instead of excluded ones
* Standardize HRRR dataloader `__getitem__` outputs to `"background"` and `"state"`
* Remove retrieval of unused ERA5 time step from HRRR dataloader
* Refactor creation of model inputs and outputs into `build_network_condition_and_target`, which is used for training, validation and inference
* Support for `invariant_array` being None, indicating no invariants
* Rename variables in inference.py, move plotting and IO to functions
* Add support for multiple rounds of gradient accumulation to enable large batch sizes
* Add support for training with AMP using float16 or bfloat16
* Align default training parameters with StormCast paper

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: examples/generative/stormcast/config/training/default.yaml
Comment (pzharrington) [inline examples/generative/stormcast/config/training/default.yaml:1]: Can we add a `lite.yaml` config that retains the short training with frequent progress/validation prints? It is useful for debugging and for a quickstart to make sure the code runs

File: examples/generative/stormcast/inference.py [hunk 1]
```diff
@@ -43,398 +45,138 @@ def main(cfg: DictConfig):
...
-            if cfg.inference.plot_var_hrrr in color_limits:
-                im = ax[0].imshow(
-                    pred[0, varidx],
-                    origin="lower",
-                    cmap="magma",
-                    clim=color_limits[cfg.inference.plot_var_hrrr],
-                )
-            else:
-                im = ax[0].imshow(pred[0, varidx], origin="lower", cmap="magma")
+            (condition, _, state_pred) = build_network_condition_and_target(
```
Comment (pzharrington) [inline examples/generative/stormcast/inference.py:126]: Let's add a comment here indicating this call is run to generate the regression model outputs, took me a bit to figure out what was going on

File: examples/generative/stormcast/inference.py [hunk 2]
```diff
@@ -43,398 +45,138 @@ def main(cfg: DictConfig):
...
+                train_regression_unet=False,
+            )
 
-            fig.colorbar(im, ax=ax[0], fraction=0.046, pad=0.04)
-            ax[0].set_title(
-                "Predicted, {}, \n initial time {} \n lead_time {} hours".format(
-                    cfg.inference.plot_var_hrrr, initial_time, i
-                )
+            state_pred_noedm = state_pred.clone()
+            edm_corrected_outputs = diffusion_model_forward(
```
Comment (pzharrington) [inline examples/generative/stormcast/inference.py:135]: And then another comment here indicating this is for the diffusion model forward pass

Comment (jleinonen) [reply examples/generative/stormcast/inference.py:137]: Comment added in [cead183](https://github.com/NVIDIA/modulus/pull/799/commits/cead1831b3831f2840cffe0503f6fd53ba827be4).

File: examples/generative/stormcast/inference.py (hunk 1 above)
Comment (jleinonen) [reply examples/generative/stormcast/inference.py:127]: Comment added in [cead183](https://github.com/NVIDIA/modulus/pull/799/commits/cead1831b3831f2840cffe0503f6fd53ba827be4).

**Review by pzharrington — APPROVED** [review]
LGTM!
